libdrgn/kdump_reader.c

```diff
@@ -0,0 +1,286 @@
+// Copyright 2019 - Serapheim Dimitropoulos
+// SPDX-License-Identifier: GPL-3.0+
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "kdump_reader.h"
+#include "program.h"
+
+bool has_kdump_signature(int fd, struct drgn_error **err)
+{
+	char signature[SIG_LEN];
+
+	int n = read(fd, signature, sizeof (signature));
+	if (n != sizeof (signature)) {
+		*err = drgn_error_format(DRGN_ERROR_FAULT,
+		                         "short read from file");
+		return false;
+	}
+
+	int cmp = memcmp(signature, KDUMP_SIGNATURE, sizeof (signature));
+	if (cmp)
+		return false;
+
+	return true;
+}
+
+struct drgn_error *drgn_kdump_init(kdump_ctx_t **ctx, int fd)
+{
+	kdump_ctx_t *context = kdump_new();
+	if (!ctx) {
+		close(fd);
```

Comment:
In general, whatever scope created a given resource should clean up that resource unless there's a clear reason to transfer ownership. Plus, it looks like this is already being closed on error by the caller.